src/main/java/com/synacy/gradprogram/spock/exercise/OrderingService.java

```diff
@@ -67,7 +72,25 @@ public Order createAnOrder(Cart cart, String recipientName, String recipientAddr
   }
 
   public void cancelOrder(CancelOrderRequest request) {
-    // TODO: Implement me. Cancels PENDING and FOR_DELIVERY orders and create a refund request saving it to the database.
-    //  Else throws an UnableToCancelException
+    String invalidOrderStatusMessage = "Cannot cancel order. Only PENDING or FOR_DELIVERY orders can be cancelled.";
+    String orderIdNotFoundMessage = "Cannot cancel order. Order ID not found.";
+
+    Optional<Order> optionalOrder = orderRepository.fetchOrderById(request.getOrderId());
+    if (optionalOrder.isPresent()) {
+
+      Order order = optionalOrder.get();
+      if (order.getStatus() != OrderStatus.PENDING && order.getStatus() != OrderStatus.FOR_DELIVERY) {
+        throw new UnableToCancelException(invalidOrderStatusMessage);
+      }
+
+      request.setDateCancelled(dateUtils.getCurrentDate());
```

Comment:
The expectation is that the request already has this value and there's no need to set it.